File: Day #284/solution.go
// Day 284: Find all cousins of a node (same level, different parent) via BFS.
// Time O(N), Space O(N).
package main

import "fmt"

type Node struct {
	val         int
	left, right *Node
}

func cousins(root *Node, target int) []int {
	q := []*Node{root}
	for len(q) > 0 {
		var next []*Node
		var level []int
		var targetParent *Node
		for _, n := range q {
			for _, c := range []*Node{n.left, n.right} {
				if c != nil {
					level = append(level, c.val)
					if c.val == target {
						targetParent = n
					}
					next = append(next, c)
				}
			}
		}
		if targetParent != nil {
			sib := make(map[int]bool)
			if targetParent.left != nil {
				sib[targetParent.left.val] = true
			}
			if targetParent.right != nil {
				sib[targetParent.right.val] = true
			}
			var res []int
			for _, v := range level {
				if !sib[v] {
					res = append(res, v)
				}
			}
			return res
		}
		q = next
	}
	return nil
}

func main() {
	root := &Node{val: 1,
		left:  &Node{val: 2, left: &Node{val: 4}, right: &Node{val: 5}},
		right: &Node{val: 3, right: &Node{val: 6}}}
	fmt.Println(cousins(root, 4)) // [6]
}

File: Day #284/solution.py
# Day 284: Find all cousins of a node (same level, different parent) via BFS.
# Time O(N), Space O(N).
from collections import deque


class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def cousins(root, target):
    q = deque([root])
    while q:
        level = []
        target_parent = None
        for _ in range(len(q)):
            n = q.popleft()
            for c in (n.left, n.right):
                if c:
                    level.append(c.val)
                    if c.val == target:
                        target_parent = n
                    q.append(c)
        if target_parent:
            sib = set()
            if target_parent.left:
                sib.add(target_parent.left.val)
            if target_parent.right:
                sib.add(target_parent.right.val)
            return [v for v in level if v not in sib]
    return []


if __name__ == "__main__":
    root = Node(1, Node(2, Node(4), Node(5)), Node(3, None, Node(6)))
    print(cousins(root, 4))  # [6]
